shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java

```diff
@@ -59,20 +63,48 @@
     }
     
     @Override
-    public TableMetaData decorate(final String tableName, final TableMetaData tableMetaData, final EncryptRule encryptRule) {
+    public Map<String, TableMetaData> decorate(final Map<String, TableMetaData> tableMetaDataMap, final EncryptRule rule, final SchemaBuilderMaterials materials) throws SQLException {
+        Map<String, Integer> dataTypes = isNeedDataTypes(tableMetaDataMap.keySet(), rule) ? getDataTypeMap(materials) : Collections.emptyMap();
+        Map<String, TableMetaData> result = new LinkedHashMap<>();
+        for (Entry<String, TableMetaData> entry : tableMetaDataMap.entrySet()) {
+            result.put(entry.getKey(), decorate(entry.getKey(), entry.getValue(), rule, dataTypes));
+        }
+        return result;
+    }
+    
+    private TableMetaData decorate(final String tableName, final TableMetaData tableMetaData, final EncryptRule encryptRule, final Map<String, Integer> dataTypeMap) {
         Optional<EncryptTable> encryptTable = encryptRule.findEncryptTable(tableName);
         return encryptTable.map(optional ->
-                new TableMetaData(tableName, getEncryptColumnMetaDataList(optional, tableMetaData.getColumns().values()), tableMetaData.getIndexes().values())).orElse(tableMetaData);
+                new TableMetaData(tableName, getEncryptColumnMetaDataList(optional, tableMetaData.getColumns().values(), dataTypeMap), tableMetaData.getIndexes().values())).orElse(tableMetaData);
+    }
+    
+    private boolean isNeedDataTypes(final Set<String> tableNames, final EncryptRule rule) {
```

Comment:
@tuichenchuxin Can we remove this for each loop, and replace this logic with public method in EncryptRule? In EncryptRule, we can store TableRule which config data type in a set collection.

Reply:
Thank you for reminding me, done.